src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,307 @@
+                ret += part;
+
+            head = end + sep.size();
+        }
+
+        return ret;
+    }
 }
 
-RequestParser::RequestParser(uint maxContentLength)
-    : m_maxContentLength(maxContentLength)
+RequestParser::RequestParser()
+    : m_request()
 {
 }
 
-RequestParser::ErrorCode RequestParser::parseHttpRequest(const QByteArray &data, Request &request)
+RequestParser::ParseResult RequestParser::parse(const QByteArray &data)
 {
-    m_request = Request();
+    // Warning! Header names are converted to lowercase
+    return RequestParser().doParse(data);
+}
 
-    // Parse HTTP request header
+RequestParser::ParseResult RequestParser::doParse(const QByteArray &data)
+{
+    // for simplicity, we don't handle malformed requests which use double `LF` as delimiter
     const int headerEnd = data.indexOf(EOH);
     if (headerEnd < 0) {
         qDebug() << Q_FUNC_INFO << "incomplete request";
-        return IncompleteRequest;
+        return {ParseStatus::Incomplete, Request(), 0};
     }
 
-    if (!parseHttpHeader(data.left(headerEnd))) {
+    const QString httpHeaders = QString::fromLatin1(data.left(headerEnd));
+    if (!parseStartLines(httpHeaders)) {
         qWarning() << Q_FUNC_INFO << "header parsing error";
-        return BadRequest;
+        return {ParseStatus::BadRequest, Request(), 0};
     }
 
-    // Parse HTTP request message
-    if (m_request.headers.contains("content-length")) {
-        int contentLength = m_request.headers["content-length"].toInt();
-        if (contentLength < 0) {
-            qWarning() << Q_FUNC_INFO << "bad request: content-length is negative";
-            return BadRequest;
-        }
+    const int headerLength = headerEnd + EOH.length();
 
-        if (contentLength > static_cast<int>(m_maxContentLength)) {
+    // handle supported methods
+    if ((m_request.method == HEADER_REQUEST_METHOD_GET) || (m_request.method == HEADER_REQUEST_METHOD_HEAD)) {
+        return {ParseStatus::OK, m_request, headerLength};
+    }
+    else if (m_request.method == HEADER_REQUEST_METHOD_POST) {
+        bool ok = false;
+        const int contentLength = m_request.headers[HEADER_CONTENT_LENGTH].toInt(&ok);
+        if (!ok || (contentLength < 0)) {
+            qWarning() << Q_FUNC_INFO << "bad request: content-length invalid";
+            return {ParseStatus::BadRequest, Request(), 0};
+        }
+        if (contentLength > MAX_REQUEST_SIZE) {
             qWarning() << Q_FUNC_INFO << "bad request: message too long";
-            return BadRequest;
+            return {ParseStatus::BadRequest, Request(), 0};
         }
 
-        QByteArray content = data.mid(headerEnd + EOH.length(), contentLength);
-        if (content.length() < contentLength) {
-            qDebug() << Q_FUNC_INFO << "incomplete request";
-            return IncompleteRequest;
-        }
+        if (contentLength > 0) {
+            const QByteArray httpBody = data.mid(headerLength, contentLength);
```

Comment:
Can we omit data copying here using QByteArray::fromRawData()?

Reply:
yes, thanks!

Reply:
No, I need to check I don't read over bound, from QByteArray::mid():
>If len is -1 (the default), or pos + len >= size(), returns a byte array containing all bytes starting at position pos until the end of the byte array.

Reply:
```c++
const int remainingLength = data.size() - headerLength;
if (remainingLength < contentLength)
    return {ParseStatus::Incomplete, Request(), 0};
if (remainingLength != contentLength) {
     return {ParseStatus::BadRequest, Request(), 0};

const QByteArray httpBody = QByteArray::fromRawData(data.constData() + headerLength, contentLength);
```

Reply:
I added a helper function `QByteArray viewMid(const QByteArray &in, const int pos, const int len = -1)` which will return a view of `in`.